cloud/azure/deploy/containerapp.go

```diff
@@ -0,0 +1,415 @@
+		return cred.Username
+	}).(pulumi.StringPtrOutput)
+
+	adminPass := creds.ApplyT(func(arg interface{}) (*string, error) {
+		cred := arg.(*containerregistry.ListRegistryCredentialsResult)
+
+		if len(cred.Passwords) == 0 || cred.Passwords[0].Value == nil {
+			return nil, fmt.Errorf("cannot retrieve container registry credentials")
+		}
+
+		return cred.Passwords[0].Value, nil
+	}).(pulumi.StringPtrOutput)
+
+	for name, eu := range args.ExecutionUnits {
+		repositoryUrl := pulumi.Sprintf("%s/%s-%s-%s", res.Registry.LoginServer, args.ProjectName, name, "azure")
+
+		image, err := image.NewImage(ctx, res.Name, &image.ImageArgs{
+			SourceImage:   eu.GetImage().GetUri(),
+			RepositoryUrl: repositoryUrl,
+			Username:      adminUser.Elem(),
+			Password:      adminPass.Elem(),
+			Server:        res.Registry.LoginServer,
+			Runtime:       runtime,
+		}, pulumi.Parent(res))
+		if err != nil {
+			return nil, errors.WithMessage(err, "function image tag "+name)
+		}
+
+		res.Apps[name], err = newContainerApp(ctx, name, &ContainerAppArgs{
+			ResourceGroupName: args.ResourceGroupName,
+			Location:          args.Location,
+			SubscriptionID:    args.SubscriptionID,
+			Registry:          res.Registry,
+			RegistryUser:      adminUser,
+			RegistryPass:      adminPass,
+			ManagedEnv:        managedEnv,
+			ImageUri:          image.URI(),
+			Env:               env,
+			ExecutionUnit:     eu,
+			ManagedIdentityID: args.ManagedIdentityID,
+		}, pulumi.Parent(res))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return res, nil
+}
+
+type ContainerAppArgs struct {
+	ResourceGroupName pulumi.StringInput
+	Location          pulumi.StringInput
+	StackID           pulumi.StringInput
+	SubscriptionID    pulumi.StringInput
+	Registry          *containerregistry.Registry
+	RegistryUser      pulumi.StringPtrInput
+	RegistryPass      pulumi.StringPtrInput
+	ManagedEnv        *app.ManagedEnvironment
+	ImageUri          pulumi.StringInput
+	Env               app.EnvironmentVarArray
+	ExecutionUnit     *deploy.ExecutionUnit
+	ManagedIdentityID pulumi.StringOutput
+}
+
+type ContainerApp struct {
+	pulumi.ResourceState
+
+	Name          string
+	Sp            *ServicePrincipal
+	App           *app.ContainerApp
+	Subscriptions map[string]*eventgrid.Topic
+}
+
+// Built in role definitions for Azure
+// See below URL for mapping
+// https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles
+var RoleDefinitions = map[string]string{
+	"KVSecretsOfficer":    "b86a8fe4-44ce-4948-aee5-eccb2c155cd7",
+	"BlobDataContrib":     "ba92f5b4-2d11-453d-a403-e96b0029c9fe",
+	"QueueDataContrib":    "974c5e8b-45b9-4653-ba55-5f855dd0fb88",
+	"EventGridDataSender": "d5a91429-5739-47e2-a06b-3470a27159e7",
+	// Access for locating resources
+	"TagContributor": "4a9ae827-6dc8-4573-8ac7-8239d42aa03f",
+}
+
+func newContainerApp(ctx *pulumi.Context, name string, args *ContainerAppArgs, opts ...pulumi.ResourceOption) (*ContainerApp, error) {
```

Comment:
Could deployment code be grouped under packages for the resource type they service, similiar to the AWS refactor, also with public`new` methods. (So they're easier to re-use some of these in external provider implementations).